### PR TITLE
feat(Makefile): Makefile target build-local build for specified OS.

### PR DESCRIPTION
**Pull request by tosone**
- Target build-local build for specified OS.
- Variable TARGETS automatic generate from cmd/* directories.
- Aligin equal sign and backlash to same column.

**Release note**:

```release-note
Target build-local build for specified OS.
Variable TARGETS automatic generate from cmd/* directories.
Align equal sign and backlash to same column.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -137,4 +148,4 @@ push: container
 
 .PHONY: clean
 clean:
-	@-rm -vrf ${OUTPUT_DIR}
+	@$(RM) -vr ${OUTPUT_DIR}
```
Comment (tosone) [inline Makefile:151]: RM is one of Makefile environments, like LDFLAGS, etc. https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html

File: Makefile [hunk 2]
```diff
@@ -26,85 +26,94 @@
 #
 
 # This repo's root import path (under GOPATH).
-ROOT := github.com/caicloud/golang-template-project
+ROOT             := github.com/caicloud/golang-template-project
 
 # Target binaries. You can build multiple binaries for a single project.
-TARGETS := admin controller
```
Comment (tosone) [inline Makefile:32]: MUST be automatic generated.

File: Makefile [hunk 3]
```diff
@@ -26,85 +26,94 @@
...
+TARGETS          := $(patsubst cmd/%,%,$(wildcard cmd/*))
 
 # Container image prefix and suffix added to targets.
 # The final built images are:
 #   $[REGISTRY]/$[IMAGE_PREFIX]$[TARGET]$[IMAGE_SUFFIX]:$[VERSION]
 # $[REGISTRY] is an item from $[REGISTRIES], $[TARGET] is an item from $[TARGETS].
-IMAGE_PREFIX ?= $(strip template-)
```
Comment (tosone) [inline Makefile:38]: All of the equal signs should be in the same column.

File: Makefile [hunk 4]
```diff
@@ -26,85 +26,94 @@
...
+.PHONY: test
 test:
 	@go test -p $(CPUS) $$(go list ./... | grep -v /vendor | grep -v /test) -coverprofile=coverage.out
 	@go tool cover -func coverage.out | tail -n 1 | awk '{ print "Total coverage: " $$3 }'
 
+.PHONY: build-local
 build-local:
 	@for target in $(TARGETS); do                                                      \
-	  go build -i -v -o $(OUTPUT_DIR)/$${target} -p $(CPUS)                            \
+	  env GOOS=$(OSName) go build -i -v -o $(OUTPUT_DIR)/$${target} -p $(CPUS)         \
```
Comment (tosone) [inline Makefile:110]:
```
env OSName=linux make build-local
```
Now, I can build Linux binary on macOS. 

We have nothing dependencies about C/C++ binding, we can just do this.

File: Makefile (hunk 1 above)
Comment (tosone) [reply Makefile:151]: Code files should be have a new line. More info: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: Makefile [hunk 5]
```diff
@@ -26,85 +26,94 @@
...
 # Container registry for base images.
-BASE_REGISTRY ?= cargo.caicloud.xyz/library
+BASE_REGISTRY    ?= cargo.caicloud.xyz/library
 
 #
 # These variables should not need tweaking.
 #
 
 # It's necessary to set this because some environments don't link sh -> bash.
-export SHELL := /bin/bash
```
Comment (tosone) [inline Makefile:52]: In my opinion, this line is useless. Default SHELL is enough.

File: Makefile (hunk 4 above)
Comment (bbbmj) [reply Makefile:110]: make build-linux is ok to build Linux binary on macOS.

File: Makefile [hunk 6]
```diff
@@ -26,85 +26,94 @@
 #
 
 # This repo's root import path (under GOPATH).
-ROOT := github.com/caicloud/golang-template-project
+ROOT             := github.com/caicloud/golang-template-project
 
 # Target binaries. You can build multiple binaries for a single project.
-TARGETS := admin controller
+TARGETS          := $(patsubst cmd/%,%,$(wildcard cmd/*))
```
Comment (bbbmj) [inline Makefile:32]: LGTM for this change

File: Makefile (hunk 1 above)
Comment (bbbmj) [reply Makefile:151]: It's ok to use `@$(RM) -vr ${OUTPUT_DIR}` or `@-rm -vrf ${OUTPUT_DIR}`

File: Makefile (hunk 4 above)
Comment (tosone) [reply Makefile:110]: Every time build the linux binary with `build-linux` in docker spend too long though i just fix one line.

`build-local` meaning build on my local env. Target `build-linux` is not useful.

File: Makefile (hunk 1 above)
Comment (tosone) [reply Makefile:151]: Yes, all of them are OK.

File: Makefile (hunk 5 above)
Comment (bbbmj) [reply Makefile:52]: It's essential to specify this in some situations,  so we set it in template to restraint all to use bash

Comment (tosone) [reply Makefile:52]: OK. Sometime bash is necessary.

File: Makefile [hunk 7]
```diff
@@ -26,85 +26,94 @@
...
-BIN_DIR := $(GOPATH)/bin
-GOLANGCI_LINT := $(BIN_DIR)/golangci-lint
+GOPATH           ?= $(shell go env GOPATH)
+BIN_DIR          := $(GOPATH)/bin
+GOLANGCI_LINT    := $(BIN_DIR)/golangci-lint
+
+ifeq ($(OS),Windows_NT)
+    OSName       ?= windows
+else
+    OSName       ?= $(shell echo $(shell uname -s) | tr '[:upper:]' '[:lower:]')
```
Comment (hanxueluo) [inline Makefile]: "echo" is unnecessary.
`$(shell uname -s | tr '[:upper:]' '[:lower:]')` is ok.

Comment (tosone) [reply Makefile]: Done.

File: Makefile [hunk 8]
```diff
@@ -26,85 +26,94 @@
 #
 
 # This repo's root import path (under GOPATH).
-ROOT := github.com/caicloud/golang-template-project
+ROOT             := github.com/caicloud/golang-template-project
```
Comment (supereagle) [inline Makefile:29]: I think this format is not necessary.

File: Makefile (hunk 4 above)
Comment (supereagle) [reply Makefile:110]: This change will make no difference, as `OSName` is the OS of current machine.

If you want to build for other OS, you can just use `env GOOS =linux make build-local`.

File: Makefile (hunk 8 above)
Comment (tosone) [reply Makefile:29]: Some individual ocd...

Usually C/C++ project like this, because lots of envs should be set.

File: Makefile (hunk 4 above)
Comment (tosone) [reply Makefile:110]: Glad to see this.

I set all of variable and toolchains as Makefile envs in my others projects, so I can see them clearly.

Simple or explicit declarations, I prefer explicit, Rust tell us.

**Pull request closed without merging**